### PR TITLE
Dev refactor critical section instruction policy

### PR DESCRIPTION
**Pull request by ouyangyu**

### CODE REVIEW

**Comment by ouyangyu** [comment]
这个pr后，可以继续重构掉：NaiveCriticalSectionInstance @lixinqi @clackhan

**Review by lixinqi — APPROVED** [review]

File: oneflow/core/vm/critical_section_instruction_policy.cpp [hunk 1]
```diff
@@ -13,38 +13,30 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-#include "oneflow/core/vm/stream.h"
+#include "oneflow/core/vm/vm_object.h"
 
 namespace oneflow {
 namespace vm {
 
-void CriticalSectionBeginPhyInstrOperand::ForEachDependence(
-    const std::function<void(vm::Dependence* compute)>& DoEach) const {
+void CriticalSectionBeginInstructionPolicy::ForEachDependence(
+    const std::function<void(Dependence* compute)>& DoEach) const {
```
Comment (lixinqi) [inline oneflow/core/vm/critical_section_instruction_policy.cpp:28]: 顺带把所有的`Dependence* compute` 改成Dependence*吧

**Review by clackhan — APPROVED** [review]

File: oneflow/core/vm/release_tensor_instruction_policy.h [hunk 2]
```diff
@@ -130,60 +130,62 @@ class SlowReleaseTensorInstructionPolicy final : public ReleaseTensorInstruction
...
+  static Maybe<vm::InstructionPolicy> VisitAsyncedLaunchedCommNet(
       DataType data_type, const std::shared_ptr<vm::EagerBlobObject>& eager_blob_object,
       const Optional<vm::Stream*>& stream) {
     return Make(data_type, eager_blob_object, stream);
   }
-  static std::unique_ptr<vm::InstructionPolicy> VisitBarrier(
+  static Maybe<vm::InstructionPolicy> VisitBarrier(
       DataType data_type, const std::shared_ptr<vm::EagerBlobObject>& eager_blob_object,
       const Optional<vm::Stream*>& stream) {
     UNIMPLEMENTED();
```
Comment (lixinqi) [inline oneflow/core/vm/release_tensor_instruction_policy.h:161]: UNIMPLEMENTED_THEN_RETURN() << "ReleaseTensor instruction not supported in Barrier stream";

File: oneflow/core/vm/release_tensor_instruction_policy.h [hunk 3]
```diff
@@ -130,60 +130,62 @@ class SlowReleaseTensorInstructionPolicy final : public ReleaseTensorInstruction
...
   }
-  static std::unique_ptr<vm::InstructionPolicy> VisitCriticalSection(
+  static Maybe<vm::InstructionPolicy> VisitCriticalSection(
       DataType data_type, const std::shared_ptr<vm::EagerBlobObject>& eager_blob_object,
       const Optional<vm::Stream*>& stream) {
     UNIMPLEMENTED();
```
Comment (lixinqi) [inline oneflow/core/vm/release_tensor_instruction_policy.h:166]: UNIMPLEMENTED_THEN_RETURN() << "ReleaseTensor instruction not supported in CriticalSection stream";

File: oneflow/core/vm/release_tensor_instruction_policy.h [hunk 4]
```diff
@@ -130,60 +130,62 @@ class SlowReleaseTensorInstructionPolicy final : public ReleaseTensorInstruction
...
   }
-  static std::unique_ptr<vm::InstructionPolicy> VisitLazyJobLauncher(
+  static Maybe<vm::InstructionPolicy> VisitLazyJobLauncher(
       DataType data_type, const std::shared_ptr<vm::EagerBlobObject>& eager_blob_object,
       const Optional<vm::Stream*>& stream) {
     UNIMPLEMENTED();
```
Comment (lixinqi) [inline oneflow/core/vm/release_tensor_instruction_policy.h:171]: UNIMPLEMENTED_THEN_RETURN() << "ReleaseTensor instruction not supported in LaunchLazyJob stream";

**Review by lixinqi — APPROVED** [review]

**Pull request merged**